spec.md

```diff
@@ -0,0 +1,954 @@
+| 16 UNAUTHENTICATED | Missing provider credentials | Request does not have valid authentication credentials for the operation | Fix the provider credentials | N |
+
+The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
+This string MAY be surfaced by CMI-Client to end users.
+
+#### `DeleteMachine`
+
+A Plugin MUST implement this RPC call if it has `DELETE_MACHINE` capability.
+This RPC will be called by the CMI-Client to deprovision a machine.
+
+- If a VM corresponding to the specified `MachineID` does not exist or the artifacts associated with the VM do not exist anymore, the Plugin MUST reply `0 OK`.
+- The plugin can OPTIONALY make use of the secrets supplied in the `Secrets` map in the `DeleteMachineRequest` to communicate with the provider.
+- This operation MUST be idempotent.
+
+```protobuf
+message DeleteMachineRequest {
+    // MachineID is the id of the machine to be deleted.
+    // It should uniquely identify the real machine in cloud-provider.
+    // This field is REQUIRED.
+    string MachineID = 1;
+
+    // Secrets is the map containing necessary credentials for cloud-provider to delete the machine.
+    // This field is OPTIONAL.
+    map<string, bytes> Secrets = 2 [(cmi_secret) = true];
+}
+
+message DeleteMachineResponse {
+    // Intentionally empty.
+}
+```
+
+##### DeleteMachine Errors
+
+If the plugin is unable to complete the DeleteMachine call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+
+| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
+|-----------|-----------|-------------|-------------------|------------|
+| 0 OK | Successful | The call was successful in deleting a VM that matches supplied deletion request. |  | N |
+| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
+| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
+| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `MachineID` and make sure that it is in the desired format and not a blank value. Exact issue to be given in `.message` | Update `MachineID` to fix issues. | N |
+| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
+| 5 NOT_FOUND |  |  |  |  |
+| 6 ALREADY_EXISTS |  |  |  |  |
+| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to delete an VM and it's required dependencies | Update requestor permissions to grant the same | N |
+| 8 RESOURCE_EXHAUSTED |  |  |  |  |
+| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |
+| 10 ABORTED |  |  |  |  |
```

Comment:
Also here, this error is possible and even likely, so it should be described how it's handled by the client. `Auto Retry Required` should be Yes.